optimade/adapters/base.py

```diff
@@ -0,0 +1,115 @@
+        self._common_converters = {
+            "json": self.entry.json,  # Return JSON serialized string, see https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson
+            "dict": self.entry.dict,  # Return Python dict, see https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict
+        }
+
+    @property
+    def entry(self):
+        """Get OPTIMADE entry"""
+        return self._entry
+
+    @entry.setter
+    def entry(self, value: dict):
+        """Set OPTIMADE entry
+        If already set, print that this can _only_ be set once.
+        """
+        if self._entry is None:
+            self._entry = self.ENTRY_RESOURCE(**value)
+        else:
+            print("entry can only be set once and is already set.")
+
+    def convert(self, format: str) -> Any:
+        """Convert OPTIMADE entry to desired format"""
+        if (
+            format not in self._type_converters
+            and format not in self._common_converters
+        ):
+            raise AttributeError(
+                f"Non-valid entry type to convert to: {format}\n"
+                f"Valid entry types: {tuple(self._type_converters.keys()) + tuple(self._common_converters.keys())}"
+            )
+
+        if self._converted.get(format, None) is None:
+            if format in self._type_converters:
+                self._converted[format] = self._type_converters[format](self.entry)
+            else:
+                self._converted[format] = self._common_converters[format]()
+
+        return self._converted[format]
+
+    def __getattr__(self, name: str) -> Any:
+        """Get converted entry or attribute from OPTIMADE entry
+        Order:
+        - Try to return converted entry if using `get_<_type_converters value>`.
+        - Try to return OPTIMADE ENTRY_RESOURCE attribute.
+        - Try to return OPTIMADE ENTRY_RESOURCE.attributes attribute.
+        - Raise AttributeError
+        """
+        # get_<entry_type>
+        if name.startswith("get_"):
+            entry_type = "_".join(name.split("_")[1:])
+            return self.convert(entry_type)
+
+        # Try returning:
+        # 1. ENTRY_RESOURCE attribute
+        # 2. ENTRY_RESOURCE.attributes attribute (an ENTRY_RESOURCE property)
+        try:
+            res = getattr(self.entry, name)
+        except AttributeError:
+            # Try to see if we are dealing with a nested attribute name, e.g., attributes.species
+            if "." in name:
+                res = None
+                nested_attributes = name.split(".")
+                try:
+                    for nested_attribute in nested_attributes:
+                        res = self.__getattr__(nested_attribute)
+                except AttributeError:
+                    pass
+                else:
+                    return res
+        else:
+            return res
+
+        try:
+            res = getattr(self.entry.attributes, name)
```

Comment:
It's just occurred to me that this ends up polluting the adapter namespace with all the methods of a pydantic model (rather than just its data). We could fix this by using something like:

```suggestion
        if name in self.entry.attributes.__field_set__:
            return getattr(self.entry.attributes, name)
```

which only looks at fields set on initializ=sation of the pydantic object (could also use just `__fields__` but that seems to include some pydantic stuff like `immutable_id`).

Reply:
`immutable_id` _is_ an OPTIMADE property.

Reply:
I think `__fields_set__` returns a set of the fields that were actually also set. So a double "set" of sorts ...
While `__fields__` returns the complete model, i.e., all available properties/attributes, not excluding even the ones that weren't set when instantiating the given model with some data set.

Reply:
oops, hopefully it was just me misreading the `__fields_set__` then, it must contain it